### PR TITLE
DEP: lib: Deprecate acceptance of float (and more) in bincount.

### LINKED ISSUE #3138: Unsafe type conversion in numpy.bincount when passing list instead of array
`np.bincount` is [supposed](http://docs.scipy.org/doc/numpy/reference/generated/numpy.bincount.html) to raise a `TypeError` if the input array-like contains elements of float or complex type. However, this is not the observed behavior when the input is an array-like (in this case, a list), rather than an array. 

``` python
>>> import numpy as np
>>> x = [1.0, 2.0, 3.5, 3.7]
>>> np.bincount(x)
array([0, 1, 1, 2])
>>> np.bincount(np.array(x))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Cannot cast array data from dtype('float64') to dtype('int64') according to the rule 'safe'
>>> np.__version__
'1.7.0'
```

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
The first argument of bincount is expected to contain integers. As noted in gh-3138, it actually accepts floats, and casts them to integers with no warnings. (It also accepts other objects that can be cast to integers, so inputs such as ["1", Fraction(5, 3)] are accepted, with fractional parts silently dropped.)

This change deprecates that behavior.  Now a deprecation warning is generated if the input cannot be safely cast to integer.

For example,

```
In [1]: import numpy as np

In [2]: np.bincount([1.4, 1.5, 3.9, 5.1])
<ipython-input-2-6d924fc4fef0>:1: DeprecationWarning: Non-integer input passed to bincount. In a future version of NumPy, this will be an error.
  np.bincount([1.4, 1.5, 3.9, 5.1])
Out[2]: array([0, 2, 0, 1, 0, 1])
```

Closes gh-3138.

### CODE REVIEW

**Review by seberg — COMMENTED** [review]
Thanks, a few smaller comments.  Could you make sure we have a test for the empty list path?  (That may or may not already exist.)

File: numpy/_core/src/multiarray/compiled_base.c [hunk 1]
```diff
@@ -130,6 +130,57 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
...
+            }
+        }
+        else {
+            /* Got an empty list. */
+            Py_DECREF(tmp1);
+        }
+    }
+
+    /* Legacy conversion. This must be changed when the deprecation expires. */
     lst = (PyArrayObject *)PyArray_ContiguousFromAny(list, NPY_INTP, 1, 1);
```
[the remark this answers is not in the document]
Comment (WarrenWeckesser) [reply numpy/_core/src/multiarray/compiled_base.c:184]: Ah, wait, I see what you're saying.  If `tmp2` succeeds, just use it below.  That doesn't result in any change of behavior.

File: numpy/_core/src/multiarray/compiled_base.c [hunk 2]
```diff
@@ -130,6 +130,57 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
...
+                                                    list, NULL, 1, 1,
+                                                    NPY_ARRAY_DEFAULT, NULL);
+        if (tmp1 == NULL) {
+            goto fail;
+        }
+        if (PyArray_SIZE(tmp1) > 0) {
+            /* The input is not empty, so convert it to NPY_INTP. */
+            PyArrayObject *tmp2 = (PyArrayObject *)PyArray_ContiguousFromAny(
+                                                    (PyObject *)tmp1,
+                                                    NPY_INTP, 1, 1);
```
[the remark this answers is not in the document]
Comment (WarrenWeckesser) [reply numpy/_core/src/multiarray/compiled_base.c:155]: FYI: currently `bincount` does not work with unsigned 64 bit integers:  https://github.com/numpy/numpy/issues/17760.  I think that is a separate issue.

[the remark this answers is not in the document]
Comment (seberg) [reply numpy/_core/src/multiarray/compiled_base.c:155]: Ah OK, hadn't realized it happily converted anything for lists, but not for arrays :).

File: numpy/_core/src/multiarray/compiled_base.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (WarrenWeckesser) [reply numpy/_core/src/multiarray/compiled_base.c]: Done.

**Comment by WarrenWeckesser** [comment]
> Could you make sure we have a test for the empty list path? (That may or may not already exist.)

Done.  There were tests for an empty array, but not for an empty list.

File: numpy/_core/src/multiarray/compiled_base.c [hunk 3]
```diff
@@ -130,9 +130,56 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
...
+        }
+        if (PyArray_SIZE(tmp1) > 0) {
+            /* The input is not empty, so convert it to NPY_INTP. */
+            lst = (PyArrayObject *)PyArray_ContiguousFromAny((PyObject *)tmp1,
+                                                             NPY_INTP, 1, 1);
+            Py_DECREF(tmp1);
+            if (lst == NULL) {
+                /* Failed converting to NPY_INTP. */
+                if (PyErr_ExceptionMatches(PyExc_TypeError)) {
+                    PyErr_Clear();
```
Comment (seberg) [inline numpy/_core/src/multiarray/compiled_base.c:159]:
```suggestion
                    PyErr_Clear();
                    /* Deprecated 2024-08-02, NumPy 2.1 */
```

File: numpy/_core/src/multiarray/compiled_base.c [hunk 4]
```diff
@@ -130,9 +130,56 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
...
+                    if (DEPRECATE("Non-integer input passed to bincount. In a "
+                                  "future version of NumPy, this will be an "
+                                  "error.") < 0) {
```
Comment (seberg) [inline numpy/_core/src/multiarray/compiled_base.c:162]:
```suggestion
                                  "error. (Deprecated NumPy 2.1)") < 0) {
```

**Review by seberg — APPROVED** [review]
Looks fine, thanks.  When the deprecation is finalized, it would be nice to clean this up a bit more (the try/except feels a bit round-about, but good enough for a deprecation).

**Comment by seberg** [comment]
Ah, it's pretty darn niche, but maybe we should add a bullet deprecation release note snippet?  Sorry for forgetting, do you want to do that?

**Comment by WarrenWeckesser** [comment]
Thanks @seberg.  I'll follow up with a release note.

File: numpy/_core/src/multiarray/compiled_base.c [hunk 5]
```diff
@@ -130,6 +130,57 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
...
+        if (PyArray_SIZE(tmp1) > 0) {
+            /* The input is not empty, so convert it to NPY_INTP. */
+            PyArrayObject *tmp2 = (PyArrayObject *)PyArray_ContiguousFromAny(
+                                                    (PyObject *)tmp1,
+                                                    NPY_INTP, 1, 1);
+            Py_DECREF(tmp1);
+            if (tmp2 == NULL) {
+                /* Failed converting to NPY_INTP. */
+                if (PyErr_ExceptionMatches(PyExc_TypeError)) {
+                    PyErr_Clear();
```
Comment (kinesiatricssxilm14) [inline numpy/_core/src/multiarray/compiled_base.c:159]: hihihi

File: numpy/_core/src/multiarray/compiled_base.c [hunk 6]
```diff
@@ -130,6 +130,57 @@ arr_bincount(PyObject *NPY_UNUSED(self), PyObject *const *args,
...
+                }
+            }
+            else {
+                /* Cast to NPY_INTP succeeded, so no deprecation warning. */
+                Py_DECREF(tmp2);
+            }
+        }
+        else {
+            /* Got an empty list. */
+            Py_DECREF(tmp1);
```
Comment (kinesiatricssxilm14) [inline numpy/_core/src/multiarray/compiled_base.c:175]: hhhhhhh

**Pull request merged**